File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    static int N, ret = 987654321;
    static boolean[] visited;
    public int solution(String begin, String target, String[] words) {
        int answer = 0;
        N = words.length;
        visited = new boolean[N];
        dfs(0, begin, words, target);
        if(ret == 987654321)
            return 0;
        return ret;
    }
    
    public void dfs(int depth, String a, String[] words, String target){
        if(a.equals(target)) {
            ret = Math.min(ret, depth);
            return;
        }
        for(int i = 0; i < N; i++){
            if(!visited[i]){
                if(check(a, words[i])){
                    visited[i] = true;
                    dfs(depth+1, words[i], words, target);
                    visited[i] = false;
                }
            }
        }
    }
    
    public boolean check(String a, String b){
        int diff = 0;
        for(int i = 0; i < a.length(); i++){
            if(a.charAt(i) != b.charAt(i))
                diff++;
        }
        if(diff == 1)
            return true;
        return false;
    }
}

File: 프로그래머스/3/43163. 단어 변환/단어 변환.cpp
#include <string>
#include <vector>
#include <queue>
#include <map>
#include <iostream>

using namespace std;
map<string,bool> visited;
queue<pair<string, int>> q;

bool isDiff(string original, string target){
    int diff = 0;
    for(int i = 0; i < original.length(); i++){
        if(original[i] != target[i])
            diff++;
    }
    if(diff == 1)
        return true;
    return false;;
}

int solution(string begin, string target, vector<string> words) {
    int answer = 0;
    q.push({begin, 0});
    
    while(q.size()){
        string cur = q.front().first;
        int cnt = q.front().second;

        q.pop();
        if(cur == target){
            answer = cnt;
            break;
        }
        for(int i = 0; i < words.size(); i++){
            if(isDiff(cur, words[i]) && !visited[words[i]]){
                visited[words[i]] = true;
                q.push({words[i], cnt + 1 });
            }
        }
    }
    return answer;
}